electrum/plugins/psbt_nostr/qt.py

```diff
@@ -0,0 +1,269 @@
+
+if TYPE_CHECKING:
+    from electrum.gui.qt import ElectrumGui
+    from electrum.gui.qt.main_window import ElectrumWindow
+
+
+NOSTR_DM = 4
+
+now = lambda: int(time.time())
+
+class QReceiveSignalObject(QObject):
+    cosigner_receive_signal = pyqtSignal(object, object, object)
+
+
+class Plugin(BasePlugin):
+
+    def __init__(self, parent, config, name):
+        BasePlugin.__init__(self, parent, config, name)
+        self._init_qt_received = False
+        self.cosigner_wallets = {}  # type: Dict[Abstract_Wallet, CosignerWallet]
+
+
+    @hook
+    def init_qt(self, gui: 'ElectrumGui'):
+        if self._init_qt_received:  # only need/want the first signal
+            return
+        self._init_qt_received = True
+        for window in gui.windows:
+            self.load_wallet(window.wallet, window)
+
+    @hook
+    def load_wallet(self, wallet: 'Abstract_Wallet', window: 'ElectrumWindow'):
+        if type(wallet) != Multisig_Wallet:
+            return
+        self.cosigner_wallets[wallet] = CosignerWallet(wallet, window)
+
+    @hook
+    def on_close_window(self, window):
+        wallet = window.wallet
+        if cw := self.cosigner_wallets.get(wallet):
+            cw.close()
+            self.cosigner_wallets.pop(wallet)
+
+    def is_available(self):
+        return True
+
+    @hook
+    def transaction_dialog(self, d: 'TxDialog'):
+        if cw := self.cosigner_wallets.get(d.wallet):
+            cw.hook_transaction_dialog(d)
+
+    @hook
+    def transaction_dialog_update(self, d: 'TxDialog'):
+        if cw := self.cosigner_wallets.get(d.wallet):
+            cw.hook_transaction_dialog_update(d)
+
+
+class CosignerWallet(Logger):
+    # one for each open window
+    # if user signs a tx, we have the password
+    # if user receives a dm? needs to enter password first
+
+    KEEP_DELAY = 24*60*60
+
+    def __init__(self, wallet: 'Multisig_Wallet', window: 'ElectrumWindow'):
+        assert isinstance(wallet, Multisig_Wallet)
+        self.wallet = wallet
+        self.network = window.network
+        self.config = self.wallet.config
+        self.window = window
+
+        Logger.__init__(self)
+        self.known_events = wallet.db.get_dict('cosigner_events')
+        for k, v in list(self.known_events.items()):
+            if v < now() - self.KEEP_DELAY:
+                self.logger.info(f'deleting old event {k}')
+                self.known_events.pop(k)
+        self.relays = self.config.NOSTR_RELAYS.split(',')
+        self.logger.info(f'relays {self.relays}')
+        self.obj = QReceiveSignalObject()
+        self.obj.cosigner_receive_signal.connect(self.on_receive)
+
+        self.cosigner_list = []  # type: List[Tuple[str, bytes, str]]
+
+        for key, keystore in wallet.keystores.items():
+            xpub = keystore.get_master_public_key()  # type: str
+            privkey = sha256('nostr_psbt:' + xpub)
+            pubkey = ecc.ECPrivkey(privkey).get_public_key_bytes()[1:]
```

Comment:
Not sure if we want to use ypub/zpubs here for a new protocol.
but I'll think about this a bit more.

Reply:
yes, good point.